aws/resource_aws_sagemaker_endpoint.go

```diff
@@ -0,0 +1,224 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/sagemaker"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsSagemakerEndpoint() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsSagemakerEndpointCreate,
+		Read:   resourceAwsSagemakerEndpointRead,
+		Update: resourceAwsSagemakerEndpointUpdate,
+		Delete: resourceAwsSagemakerEndpointDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"name": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
+				ForceNew:     true,
+				ValidateFunc: validateSagemakerName,
+			},
+
+			"endpoint_config_name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validateSagemakerName,
+			},
+
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceAwsSagemakerEndpointCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).sagemakerconn
+
+	var name string
+	if v, ok := d.GetOk("name"); ok {
+		name = v.(string)
+	} else {
+		name = resource.UniqueId()
+	}
+
+	createOpts := &sagemaker.CreateEndpointInput{
+		EndpointName:       aws.String(name),
+		EndpointConfigName: aws.String(d.Get("endpoint_config_name").(string)),
+	}
+
+	if v, ok := d.GetOk("tags"); ok {
+		createOpts.Tags = tagsFromMapSagemaker(v.(map[string]interface{}))
+	}
+
+	log.Printf("[DEBUG] SageMaker endpoint create config: %#v", *createOpts)
+	_, err := conn.CreateEndpoint(createOpts)
+	if err != nil {
+		return fmt.Errorf("error creating SageMaker endpoint: %s", err)
+	}
+
+	d.SetId(name)
+	log.Printf("[INFO] SageMaker endpoint ID: %s", d.Id())
+
+	log.Printf("[DEBUG] Waiting for SageMaker endpoint (%s) to become available", d.Id())
+	stateConf := &resource.StateChangeConf{
```

Comment:
It may be worth trying to use the AWS Go SDK provided waiter functions ([`WaitUntilEndpointDeleted()`](https://docs.aws.amazon.com/sdk-for-go/api/service/sagemaker/#SageMaker.WaitUntilEndpointDeleted) / [`WaitUntilEndpointInService()`](https://docs.aws.amazon.com/sdk-for-go/api/service/sagemaker/#SageMaker.WaitUntilEndpointInService)) here, during update, and during deletion to simplify our code instead of using `StateChangeConf`, e.g.

```go
	_, err := conn.CreateEndpoint(createOpts)

	if err != nil {
		return fmt.Errorf("error creating SageMaker Endpoint: %s", err)
	}

	d.SetId(name)

	describeInput := &sagemaker.DescribeEndpointInput{
		EndpointName: aws.String(name),
	}

	if err := conn.WaitUntilEndpointInService(describeInput); err != nil {
		return fmt.Errorf("error waiting for SageMaker Endpoint (%s) to be in service: %s", name, err)
	}

	return resourceAwsSagemakerEndpointRead(d, meta)
```